[JENKINS-26445] Build history pagination and search

tfennelly opened the pull request:
Currently, Jenkins tries to load every build into the Build History list if the user presses the "More" link, resulting in a suckie experience if there are 1000s of builds.

[See JENKINS-26445](https://issues.jenkins-ci.org/browse/JENKINS-26445)

This change set:
1. Adds pagination by spliting the history into consumable "pages".
2. There are page up, page down and home controls for navigating the pages.
3. There's a find/search box for narrowing the list of builds being displayed.

It supports a "fuzzy" matcher.

![screenshot 2015-09-16 12 15 14](https://cloud.githubusercontent.com/assets/429311/9903297/465525a2-5c6d-11e5-9778-ab3d18dda129.png)

Would be good to add some nicer icons, but we can do that in another PR. Quite a few people have been asking for this so it would be good to get it in there if we can.

tfennelly committed:
Build history pagination and search

tfennelly commented:
@reviewbybees

KostyaSha commented:
I don't understand why everybody stick to this small menu, why not just create powerful build history page for project and use full main_panel area with sorting and etc?

daniel-beck commented:
Given the scope of changes to the sidepanel I agree with @KostyaSha. A dedicated builds page would probably be better than overloading the sidepanel with features.

Of course this doesn't mean filter/search, pagination etc. are useless; they'd make great features for a dedicated builds page.

(The ItemGroup-level builds page needs some love as well -- maybe these can be merged?)

KostyaSha commented on core/src/main/java/hudson/widgets/HistoryPageEntry.java:
```diff
@@ -0,0 +1,67 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2013-2014, CloudBees, Inc.
```
2015 already :)

tfennelly commented:
@KostyaSha @daniel-beck Your question re why the Build History widget is confined to the left side panel is has surely been a long-standing question (long before now), no? In any case, I agree with you on that. However, I don't see how these changes have made that any worse. If anything, I would think it makes the widget more useable in the side panel + nothing has been done here that precludes us from adding an "expand to full page" control on it in another PR.

Anyway, we can keep this PR open for another while and see if there is any interest in it. If not, then it's obviously not as big an issue as I had thought, so no problem with scrapping these changes (or putting them on ice somewhere).

arpitgupta commented:
I also agree with Tom. That these changes are helpful in the current form. If we decide to implement a build history page we can certainly start using that.

The workaround that we had to implement in order to get around this issue is  build a tool that given a job name would find the different builds in jenkins that had that name. As the current code freezes the browser and makes it un usable. The other workaround we did we only keep builds for 1 day. Before the re design we were keeping the build for a lot longer.

So i am certainly +1 on adding these changes in to jenkins and we can create new jira for doing a build history page.

daniel-beck commented:
This is an improvement over the current situation, no doubt about it. Being able to search for builds looks very interesting as well (although I wonder whether this could be integrated into the regular search UI as well, to not create two similar but completely different search systems).

That said, with the added features it becomes apparent now that the builds history needs more space than it gets in the side panel. Things like showing full(er) build descriptions could then be implemented as well. A separate build timeline page already exists after all, but it shows very specific timing information and nothing else. The ItemGroup-wide builds list is weird at best. Neither is extensible. So a larger effort to clean these up would be great.

I wonder now whether it would be a good idea to first merge this, once done, then later consider moving most of its functionality to a separate page, or whether this would lock us into this design. I think it probably is the former, even though it will piss off users who get used to it too quickly.

oleg-nenashev commented on core/src/main/java/hudson/model/Queue.java:
```diff
@@ -2337,6 +2337,33 @@ boolean leave(Queue q) {
+     * No other clean way to effectively mock Item.
+     */
+    @SuppressWarnings("unchecked")
+    public static final class MockItem extends Item {
```
It makes sense to move this code to test classes. No need to keep it in core/src/main IMO

tfennelly replied:
You're right. Not sure why I thought it couldn't go into the tests. Thanks.

oleg-nenashev commented on core/src/main/java/hudson/widgets/HistoryPageEntry.java:
```diff
@@ -0,0 +1,67 @@
+
+/**
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ */
```
Javadoc could be improved

tfennelly replied:
Sure. Will do. Thanks.

oleg-nenashev commented on core/src/main/java/hudson/widgets/HistoryPageFilter.java:
```diff
@@ -0,0 +1,280 @@
+ *
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ */
+public class HistoryPageFilter<T> {
```
It could be better to have several subclasses with the logic (e.g. NewerThan, OlderThan). Merging of the logic into a single class is not good

tfennelly replied:
Sorry, don't agree with you on this one. IMO, keeping the logic together makes it easier to understand how the pagination works Vs needlessly splitting it across multiple classes.

oleg-nenashev commented on core/src/main/java/hudson/widgets/HistoryWidget.java:
```diff
@@ -117,27 +129,61 @@ public String getFirstTransientBuildKey() {
+    /**
+     * Render a "page" of records.
+     */
+    public HistoryPageFilter getPage() {
```
Rename to <code>getHistoryPageFilter()</code>?

tfennelly replied:
Sure.

oleg-nenashev commented on core/src/main/java/hudson/widgets/HistoryWidget.java:
```diff
@@ -214,4 +262,16 @@ public void setNextBuildNumberToFetch(String nextBuildNumberToFetch) {
         String getNextKey(String key);
     }
+
+    private Long getPagingParam(StaplerRequest currentRequest, String name) {
```
<code>@CheckForNull</code> is required

tfennelly replied:
I don't really think it needs it since it's an internal method, but will add it.

oleg-nenashev commented on core/src/main/java/hudson/widgets/buildsearch/BuildSearchParamProcessorFactory.java:
```diff
@@ -0,0 +1,109 @@
+     *
+     * @return List of search terms.
+     */
+    public abstract String[] getSearchTerms();
```
What about Enum or another wrapper class? Current implementation leads to hardcodes in your code

tfennelly replied:
Will look at that. There's no duplicates at the moment.

oleg-nenashev commented on core/src/main/java/hudson/widgets/buildsearch/BuildSearchParams.java:
```diff
@@ -0,0 +1,174 @@
+        return searchParamsSets.toString();
+    }
+
+    private void parseSearchString() {
```
I suppose the manual search implementations here could be replaced by common methods available in Apache Commons

tfennelly replied:
No idea. If you have something specific in mind then let me know. Thanks.

oleg-nenashev commented:
I would vote for the change, but :+1: for the separate page with appropriate extension points and API/CLI support

tfennelly committed:
Fixed screwy indent

tfennelly commented:
@oleg-nenashev thanks for the review and comments. I'll make fixes to the comments we agree on.

As for the discussion on a separate page ( @daniel-beck @KostyaSha ). If you guys believe strongly enough in that then feel free to create a JIRA for it and maybe someone will pick it up.

tfennelly committed:
Moved Mock Queue.Item out into test/java

Not sure why I ever had it in Queue.MockItem. I must have wrongly though it was private/protected in some way that did not allow it to live anywhere else.

jglick commented on core/src/main/java/hudson/widgets/buildsearch/BuildSearchParamProcessor.java:
```diff
@@ -0,0 +1,91 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package hudson.widgets.buildsearch;
```
New types should go into subpackages of `jenkins`.

jglick commented on core/src/main/java/hudson/widgets/buildsearch/BuildSearchParamProcessorFactory.java:
```diff
@@ -0,0 +1,109 @@
+
+    static {
+        // Statically adding the core BuildSearchParamProcessorFactory impls Vs adding via @Extension.
+        // Makes testing easier coz no need to create a JenkinsRule instance + put in test harness etc.
```
Better to define even these via `@Extension`. You do not need `JenkinsRule` to call `ExtensionList.lookup`: from a unit test it will look for implementations in the classpath.

tfennelly replied:
> You do not need JenkinsRule to call ExtensionList.lookup: from a unit test it will look for implementations in the classpath.

Fairly sure I tried this and got an empty list. Will try again and see.

jglick commented:
This seems like a big block of new code. IMO it would be better moved into its own plugin. I filed [JENKINS-27919](https://issues.jenkins-ci.org/browse/JENKINS-27919) to track that.

tfennelly commented:
> This seems like a big block of new code. IMO it would be better moved into its own plugin. I filed JENKINS-27919 to track that.

:sob:

KostyaSha commented:
My proposal about new page was exactly because UI for this small widget and its code becomes over complex. IMO providing link for search from this widget to good page is something that i expect.

tfennelly committed:
Build History pagination/search tweaks based on feedback from @oleg-nenashev

tfennelly committed:
Moved Build History pagination and search classes to jenkins as requested by @jglick

I understand the motivation for this but I think it creates a bit of a maint/comprehension issue in that splits related classes across 2 different package hierarchies. I know we'll say that's ok but I don't like it. Just say'n.

tfennelly commented:
@oleg-nenashev @jglick eventually got back to acting on your comments on this PR.

@jglick you said `ExtensionList.lookup` didn't need `JenkinsRule` and would pick up `@Extension` impls from the classpath. [I tried that again](https://github.com/tfennelly/jenkins/commit/894ec7d396fc6d1d589402dadf3252a184702502) and it didn't work for me without a `JenkinsRule`. You think that should work for sure?

jglick commented:
> You think that should work for sure?

I had thought so, but no [I just thought it](https://github.com/jenkinsci/workflow-plugin/blob/d0065d2cd492fa1c722af6ff3b0f440c45707b07/step-api/src/main/java/org/jenkinsci/plugins/workflow/structs/DescribableHelper.java#L430-448). You could [fix that yourself](https://github.com/jglick/jenkins/blob/cd8dbedaf0f8385a13da543b0c8e0ec72d3e53c7/core/src/main/java/hudson/ExtensionList.java#L128).

tfennelly committed:
Fix jelly refs to historyPageFilter

tfennelly committed:
Moved some jelly files to match earlier Java package move

gusreiber commented:
For those suggesting this should be its own page, does that mean we should remove the history widget on the left all together, or that we should leave it in its somewhat broken state? It seems to me that removing it from the left will generate an outcry of complaints who get value from the current feature.

tfennelly committed:
Merge branch 'master' into JENKINS-26445-pagination-search

* master: (256 commits)
  Noting #1817
  Noting #1825
  updated changelog for release
  [maven-release-plugin] prepare for next development iteration
  [maven-release-plugin] prepare release jenkins-1.629
  [JENKINS-30139] Oleg wants to keep the data accessible so we have to go for immutable
  [JENKINS-30139] Forgot to change this one
  [JENKINS-30139] Add @Restricted to prevent external use of internal data
  [JENKINS-30139] Switch from a big fat lock to a concurrent collection
  [JENKINS-30374] Update obsolete help message
  updated changelog for release
  [maven-release-plugin] prepare for next development iteration
  [maven-release-plugin] prepare release jenkins-1.628
  noting merge of PR#1816 - use slf4j interceptors
  Manage commons-logging and log4j as provided, include slf4j adapters for tests.
  Exclude commons-logging and log4j in dangerous scopes compile and runtime in plugins as well.
  Only exclude commons-logging and log4j in dangerous scopes compile and runtime.
  Reintroduce enforcer rules as suggested by @jtnord.
  Document allBuilds for Job remote access API
  Include slf4j-jdk14 during tests.
  ...

Conflicts:
	war/src/main/webapp/scripts/hudson-behavior.js

tfennelly committed:
Added a fuzzy search for when search terms are not used

Easier for users - nothing to learn

tfennelly commented:
Flaky build failure (it seems). Closing and reopening to test.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

rsandell commented:
For future consideration it would be nice if the searchbox has some pluggability, so that other plugins could partake in the filtering, for example https://wiki.jenkins-ci.org/display/JENKINS/Metadata+plugin

tfennelly commented:
@rsandell That's there in this PR actually. See the `BuildSearchParamProcessorFactory` `@Extension`point.

rsandell commented:
Well that's cool. I only skimmed through it, but I'm not fully sure my example would fit since the metadata plugin has its own expression compiler etc.

Anyway my example is a bit bad since that plugin is still in beta and hasn't been actively developed on in years :)

tfennelly commented:
@rsandell yeah, I'm not convinced that extension will be used but, in the spirit of Jenkins OTT extensibility, it's there ;)

jglick commented:
> I'm not convinced that extension will be used but, in the spirit of Jenkins OTT extensibility, it's there

There should not be an extension point without a well-defined use case and at least one (preferably two or three) sample implementations.

tfennelly commented:
> There should not be an extension point without a well-defined use case and at least one (preferably two or three) sample implementations.

There are 4 or 5 implementations of it.

oleg-nenashev commented:
Cannot be merged now

tfennelly commented:
@oleg-nenashev there's no plan to merge it atm. I'll fix it up before we try. @daniel-beck wants to remove some functionality from it anyway.

tfennelly committed:
Merge branch 'master' into JENKINS-26445-pagination-search

tfennelly committed:
Removed more advanced search options as requested by @daniel-beck

tfennelly commented:
I removed the fine grained search tool. Now just has the "fuzzy" matcher. Hopefully @daniel-beck will be happy now :)

The code for the more advanced search is on https://github.com/tfennelly/jenkins/tree/pagination-search-advanced

tfennelly committed:
Disable build history refresh if the pane is collapsed

See JENKINS-30569 and https://github.com/jenkinsci/jenkins/commit/6cadf70d3afb1d1786bf02e549baea99d3ab308a

daniel-beck commented:
Thanks a lot @tfennelly the result is pretty awesome!

batmat commented:
Wow +1 :+1: congrats & thanks @tfennelly that one is really gonna improve user experience.

KostyaSha commented on core/src/main/java/hudson/widgets/HistoryWidget.java:
```diff
@@ -83,10 +88,14 @@
      *      The parent model object that owns this widget.
      */
     public HistoryWidget(O owner, Iterable<T> baseList, Adapter<? super T> adapter) {
+	StaplerRequest currentRequest = Stapler.getCurrentRequest();
```
intend?

KostyaSha commented on core/src/main/java/hudson/widgets/HistoryWidget.java:
```diff
@@ -117,29 +128,63 @@ public String getFirstTransientBuildKey() {
         } else {
-            // to prevent baseList's concrete type from getting picked up by <j:forEach> in view
-            return updateFirstTransientBuildKey(Iterators.wrap(baseList));
+                // to prevent baseList's concrete type from getting picked up by <j:forEach> in view
```
indent?
